kairon/shared/cognition/processor.py

```diff
@@ -0,0 +1,258 @@
+
+    def is_collection_limit_exceeded(self, bot, collection):
+        """
+        checks if collection limit is exhausted
+
+        :param bot: bot id
+        :param collection: Name of collection
+        :return: boolean
+        :raises: AppException
+        """
+
+        collections = list(CognitionSchema.objects(bot=bot).distinct(field='collection_name'))
+        if collection not in collections and len(collections) >= BotSettings.objects(
+                bot=bot).get().cognition_collections_limit:
+            return True
+        else:
+            return False
+
+    def is_column_collection_limit_exceeded(self, bot, metadata):
+        """
+        checks if columns in collection limit is exhausted
+
+        :param bot: bot id
+        :param metadata: schema
+        :return: boolean
+        """
+        return len(metadata) >= BotSettings.objects(bot=bot).get().cognition_columns_per_collection_limit
+
+    def is_same_column_in_metadata(self, metadata):
+        """
+        checks if there are same columns in metadata
+
+        :param bot: bot id
+        :param metadata: schema
+        :return: boolean
+        """
+        if len(metadata) < 2:
+            return False
+        reference_value = metadata[0].get('column_name')
+        return all(d.get('column_name') == reference_value for d in metadata)
+
+    def get_content(self, bot: Text, **kwargs):
+        """
+        fetches content
+
+        :param bot: bot id
+        :return: yield dict
+        """
+        kwargs["bot"] = bot
+        search = kwargs.pop('data', None)
+        start_idx = kwargs.pop('start_idx', 0)
+        page_size = kwargs.pop('page_size', 10)
+        cognition_data = CognitionData.objects(**kwargs)
+        if search:
+            cognition_data = cognition_data.search_text(search)
+        for value in cognition_data.skip(start_idx).limit(page_size):
+            item = value.to_mongo().to_dict()
+            item.pop('timestamp')
+            item["_id"] = item["_id"].__str__()
+            yield item
+
+    def list_cognition_collections(self, bot: Text):
+        """
+        Retrieve cognition data.
+        :param bot: bot id
+        """
+        collections = list(CognitionData.objects(bot=bot).distinct(field='collection'))
+        return collections
+
+    def save_cognition_schema(self, metadata: Dict, user: Text, bot: Text):
+        if self.is_collection_limit_exceeded(bot, metadata.get('collection_name')):
+            raise AppException('Collection limit exceeded!')
+        if metadata.get('metadata') and self.is_column_collection_limit_exceeded(bot, metadata.get('metadata')):
+            raise AppException('Column limit exceeded for collection!')
+        if metadata.get('metadata') and self.is_same_column_in_metadata(metadata.get('metadata')):
+            raise AppException('Columns cannot be same in the schema!')
+        metadata_obj = CognitionSchema(bot=bot, user=user)
+        if metadata.get('metadata'):
```

Comment:
remove if condition
metadata_obj.metadata = [ColumnMetadata(**meta) for meta in metadata.get('metadata') or []]